src/bridge/crypto/rsa.cc

```diff
@@ -0,0 +1,250 @@
+// specification does not explicitly permit this function to be null in an
+// engine's `RSA_METHOD` implementation, so we define it as a no-op here.
+//
+// Returns 1 on success; otherwise, returns 0.
+int Init(RSA *rsa) {
+  return true;
+}
+
+// Cleans up any internal structures associated with the input `rsa` struct
+// (except for the RSA struct itself, which will be cleaned up by the OpenSSL
+// library).
+//
+// Called when OpenSSL's `RSA_free` is called on `rsa`. (See the OpenSSL man
+// page for RSA_free(3) for more information.)
+//
+// Returns 1 on success; otherwise, returns 0.
+int Finish(RSA *rsa) {
+  // `crypto_key_handle` is guaranteed to be non-null here (if the underlying
+  // external data struct was null, an error status would be returned).
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const CryptoKeyHandle *crypto_key_handle,
+      GetCryptoKeyHandleFromOpenSslRsa(rsa), false);
+  delete crypto_key_handle;
+
+  Status status = AttachCryptoKeyHandleToOpenSslRsa(nullptr, rsa);
+  if (!status.ok()) {
+    KMSENGINE_SIGNAL_ERROR(status);
+  }
+  return status.ok();
+}
+
+// Signs the message digest `digest_bytes` of length `digest_length` using the
+// RSA private key represented by the OpenSSL RSA struct `rsa`. Then, stores the
+// resulting signature in `signature_return` and the signature size in
+// `signature_length`. The caller is responsible for ensuring that
+// `signature_return` points to `RSA_size(rsa)` bytes of memory. (See the
+// OpenSSL man page for RSA_sign(3) for more information.)
+//
+// Returns 1 on success; otherwise, returns 0.
+//
+// The function signature comes from the prototype for `RSA_meth_set_sign` from
+// the OpenSSL API.
+int Sign(int type, const unsigned char *digest_bytes,
+         unsigned int digest_length, unsigned char *signature_return,
+         unsigned int *signature_length, const RSA *rsa) {
+  // Convert arguments to engine-native structures for convenience. These
+  // conversions need to take place within the bridge layer (as opposed to
+  // letting the `RsaKey::Sign` method handling the conversions) since the
+  // conversion functions refer to some OpenSSL API functions.
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const CryptoKeyHandle *crypto_key_handle,
+      GetCryptoKeyHandleFromOpenSslRsa(rsa), false);
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const DigestCase digest_type,
+      ConvertOpenSslNidToDigestType(type), false);
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const std::string digest_string,
+      MakeDigestString(digest_bytes, digest_length), false);
+
+  // Delegate handling of the signing operation to the backing layer.
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const std::string signature,
+      crypto_key_handle->Sign(digest_type, digest_string), false);
+
+  // Copy results into the return pointers.
+  if (signature_return != nullptr) {
```

Comment:
I think the calling code is responsible to check the `RSA_size` requirement according to OpenSSL, but this wouldn't hurt to check, so I'll add it. If that was false, then some precondition failed anyways and the user should know.

Reply:
I think checking `signature.length() <= RSA_size(…)` makes sense here. I think the `RSA_size > 0` check is unnecessary since calling code must call that in order to know how big of a buffer to allocate for the signature prior to this (and the `>0` is already captured in the comparison against `.length()`).

Reply:
Actually, it looks like this will break the current test cases because `RSA_size` requires certain fields of the `RSA` struct to be populated (which are currently not being populated). We can write test cases for this though pending #93, so I'll hold on this for a bit.

Reply:
Okay, #93 is merged in so this is good to go. See test case for updated strategy to populate a well-formed `RSA` struct.